[PodLevelResources] Pod Level Resources Feature Alpha

ndixita opened the pull request:
#### What type of PR is this?

<!--
Add one of the following kinds:
/kind feature
/kind api-change

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR implements Pod Level Resources that require following changes:
1. API changes to support Resources in PodSpec
2. Kubectl changes to display Resources at pod-level
3. Defaulting logic when pod-level requests are missing but limits are set.
4. Validation logic for pod-level resources 
5. Scheduler changes to consume pod-level requests
6. QoS class determination changes
7. cgroup changes to use pod-level resources
8. OOM score adjustment calculation changes to account for pod-level requests.
9. e2e tests
10. Limit Range changes for type Pod to check against pod-level resources, if set.
11. Resource Quota changes to check against pod-level resource, if set.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #
xref: https://github.com/kubernetes/enhancements/issues/2837

#### Special notes for your reviewer:
API changes: https://github.com/kubernetes/kubernetes/pull/128407/commits/62c535517655a9133513db0bdb5dccc581338ee3

Scheduler changes: https://github.com/kubernetes/kubernetes/pull/128407/commits/cb4f2ea7177e6cb248136599579a8a1796ef677f 

Kubelet changes: inside pkg/kubelet


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
- Changed the Pod API to support `resources` at `spec` level for pod-level resources.

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2837-pod-level-resource-spec/README.md 
```

ndixita committed:
API changes for Pod Level Resources
1. Add Resources struct to PodSpec struct in both external and internal API packages
2. Adding feature gate and logic for dropping disabled fields for Pod Level Resources
KEP: enhancements/keps/sig-node/2837-pod-level-resource-spec

ndixita committed:
Generated files and compatability data from API changes

ndixita committed:
Adding support for pod level resources in kubectl
1. Add support for pod level resources in kubectl
2. Reuse the existing method to describe container resources and generalize it to describe both pod and container level resources

ndixita committed:
Adding the logic to set default pod-level request as following:
1. If pod-level limit is set, pod-level request is unset and container-level request is set: derive pod-level request from container-level requests
2. If pod-level limit is set, pod-level request is unset and container-level request is unset: set pod-level request equal to pod-level limit

ndixita committed:
Adding the logic to validate pod-level resources as following:
1. The effective container requests cannot be greater than pod-level requests
2. Inidividual container limits cannot be greater than pod-level limits
3. Only CPU & Memory are supported at pod-level
4. Inplace container resources updates are not supported if pod-level resources are set
Note: effective container requests cannot be greater than pod-level limits is supported by transitivity. Effective container requests <= pod-level requests && pod-level requests <= pod-level limits; Therefore effective container requests <= pod-level limits

Signed-off-by: ndixita <ndixita@google.com>

ndixita committed:
Scheduler changes:
1. Use pod-level resource when feature is enabled and resources are set at pod-level
2. Edge case handling: When a pod defines only CPU or memory limits at pod-level (but not both), and container-level requests/limits are unset, the pod-level requests stay empty for the resource without a pod-limit. The container's request for that resource is then set to the default request value from schedutil.

ndixita commented:
/test all

ndixita committed:
QOS changes for Pod Level resources

ndixita committed:
cgroup configuration changes:
1. Pod cgrooup configured to use resources from pod spec if feature is enabled and resources are set at pod-level
2. Container cgroup limits defaulted to pod-level limits is container limits are not set

ndixita commented:
/test all

kannon92 commented:
Hey @ndixita, which PR are you working on? I see https://github.com/kubernetes/kubernetes/pull/128406 and this one.

ndixita commented:
/test all

ndixita commented:
Addressing the default test failures

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4184,100 +4186,166 @@ func validateHostIPs(pod *core.Pod) field.ErrorList {
 // The pod metadata is needed to validate generic ephemeral volumes. It is optional
 // and should be left empty unless the spec is from a real pod object.
-func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
+func ValidatePodSpec(pod *core.Pod, podMeta *metav1.ObjectMeta, fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
```
rather than changing the first arg to *core.Pod, if pod status is *conditionally* needed to do some aspects of validation, making that an optional secondary arg might be clearer (just like how podMeta is an optional secondary arg currently)

ndixita replied:
Fixed

yujuhong commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -18619,6 +18619,54 @@ func TestValidateServiceUpdate(t *testing.T) {
+	for k, item := range table {
+		err := validatePodResourceName(item.input, field.NewPath("field"))
+		if len(err) != 0 && item.success {
+			t.Errorf("expected no failure for input %q", item.input)
```
print out the `err`

yujuhong commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -18619,6 +18619,54 @@ func TestValidateServiceUpdate(t *testing.T) {
+		{"kubernetes.io/will/not/work/", false, ""},
+	}
+	for k, item := range table {
+		err := validatePodResourceName(item.input, field.NewPath("field"))
```
nit: s/err/errs

yujuhong commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -18619,6 +18619,54 @@ func TestValidateServiceUpdate(t *testing.T) {
+			t.Errorf("expected no failure for input %q", item.input)
+		} else if len(err) == 0 && !item.success {
+			t.Errorf("expected failure for input %q", item.input)
+			for i := range err {
```
This should be for the case where `!item.success` and `len(err)!=0`

yujuhong commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -18619,6 +18619,54 @@ func TestValidateServiceUpdate(t *testing.T) {
+	table := []struct {
+		input   core.ResourceName
+		success bool
+		expect  string
```
I don't see `expect` being used in any of the test case.

ndixita replied:
Fixed, had copied the test code from the function below this function (which requires refactoring).

yujuhong commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6348,10 +6429,48 @@ func validateContainerResourceName(value core.ResourceName, fldPath *field.Path)
+func getResourceNameWithoutPrefix(prefixedResource core.ResourceName) (string, error) {
+	segments := strings.Split(string(prefixedResource), "/")
+	if len(segments) > 2 {
+		return "", fmt.Errorf("must be a supported resource for pod-level resources")
```
Is this function supposed to be general purpose, or only for validation pod resource? The function name suggests the former, but the error message the latter.

yujuhong commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4184,100 +4186,166 @@ func validateHostIPs(pod *core.Pod) field.ErrorList {
+		AllowIndivisibleHugePagesValues: opts.AllowIndivisibleHugePagesValues,
+		PodLevelResource:                true,
+	}
+	// ValidateResourceRequirements ensures
```
unfinished comment

ndixita replied:
Fixed

yujuhong commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4184,100 +4186,166 @@ func validateHostIPs(pod *core.Pod) field.ErrorList {
+	return allErrs
+}
+
+func validatePodResourceConsistency(pod *core.Pod, fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
```
Is there a separate unit test for this?

ndixita replied:
Added the tests

yujuhong commented:
@tallclair for review as well

ndixita commented:
/test all

ndixita commented:
/assign @tallclair

ndixita commented:
/assign @AxeZhan

cici37 commented:
/remove-sig api-machinery

ndixita committed:
Adding OOM Score adjustment formula changes that takes pod level resources into account

Signed-off-by: ndixita <ndixita@google.com>

ndixita commented on pkg/apis/core/helper/qos/qos.go:
```diff
@@ -39,6 +41,45 @@ func GetPodQOS(pod *core.Pod) core.PodQOSClass {
 	return ComputePodQOS(pod)
 }
 
+// zeroQuantity represents a resource.Quantity with value "0", used as a baseline
```
No idea! But https://github.com/kubernetes/kubernetes/blob/365b457e3eb1637f3d4c3d6b8815ee554aa36a24/pkg/apis/core/v1/helper/qos/qos_test.go#L139 covers testing this package.

yujuhong replied:
It's weird that this doesn't get the same coverage in both packages, but okay with not changing this unless other reviewers have opinions.

ndixita commented:
/assign @roycaihw

ndixita commented:
/assign @haircommander

haircommander commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -44,15 +45,80 @@ type PodResourcesOptions struct {
+
+// IsPodLevelResourcesSet check if PodLevelResources feature is enabled and
+// pod-level requests are set.
+func isPodLevelRequestsSet(pod *v1.Pod, podLevelResourcesenabled bool) bool {
```
nit: `podLevelResourcesEnabled`*

haircommander commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -379,6 +379,571 @@ func testPodDefaults(t *testing.T, featuresEnabled bool) {
+			Containers: []v1.Container{{Resources: v1.ResourceRequirements{}}, {Resources: v1.ResourceRequirements{}}},
+		},
+	}, {
+		name: "pod resources=unset, container requests=unset limits=set",
```
nit: `unset, limits=set`

haircommander replied:
or maybe `;` as it's not acting on the noun container (i.e: it's pod + container limits set, not `pod resources=set; container requests=unset, limits=set`

haircommander replied:
oh I see it's acting on the noun container. nevermind

haircommander commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -379,6 +379,571 @@ func testPodDefaults(t *testing.T, featuresEnabled bool) {
+					"memory": resource.MustParse("8Mi"),
+				},
+			},
+			Containers: []v1.Container{
```
from what I can tell this is the only test that tests a non homogeneously set group of containers (where they have different permutations of requests/limits set). I am curious about tests with two containers A and B:
A limits !request, B requests !limit
A limits request, B requests !limit
A limits request, B !requests limit // this is covered here
with and without all pod resources (limits, requests, limits & requests)

ndixita replied:
Tests for these cases are explicitly not needed because unit tests of PodRequests covers all these cases. Whether 1 container or multiple containers have requests/limits set/unset, the defaulting logic just uses the calculation PodRequests returns. I will still add a few more test cases in the follow up PR.

haircommander commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -429,3 +435,55 @@ func SetDefaults_HostPathVolumeSource(obj *v1.HostPathVolumeSource) {
+	podReqsFromContainers := resourcehelper.PodRequests(obj, resourcehelper.PodResourcesOptions{
+		InPlacePodVerticalScalingEnabled: utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling),
+		PodLevelResourcesEnabled:         utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources),
+		UseContainerLevelResources:       true,
```
I'm confused, why do we have this field? will it be an additional feature gate or something?

haircommander commented on pkg/kubelet/cm/helpers_linux.go:
```diff
@@ -119,16 +119,19 @@ func HugePageLimits(resourceList v1.ResourceList) map[int64]int64 {
 	reqs := resource.PodRequests(pod, resource.PodResourcesOptions{
 		InPlacePodVerticalScalingEnabled: inPlacePodVerticalScalingEnabled,
+		PodLevelResourcesEnabled:         podLevelResourcesEnabled,
 	})
```
if `UseContainerLevelResources` should be false here then I think it's worth adding it to the struct here and adding a comment why

ndixita replied:
removed the field.

cici37 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6695,8 +6814,14 @@ func validateBasicResource(quantity resource.Quantity, fldPath *field.Path) fiel
+// ResourceValidationOptions
+type ResourceValidationOptions struct {
+	PodLevelResource                bool
+	AllowIndivisibleHugePagesValues bool
```
Worth documenting the opt added VS PodValidationOptions.AllowIndivisibleHugePagesValues. Looks like they always share the same value?

ndixita replied:
removed the ResourceValidationOptions struct and refactored the code.

ndixita commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -429,3 +435,55 @@ func SetDefaults_HostPathVolumeSource(obj *v1.HostPathVolumeSource) {
+// while mainting compatability with the container-level specifications, as detailed
+// in KEP-2837: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2837-pod-level-resource-spec/README.md#proposed-validation--defaulting-rules
+func defaultPodResources(obj *v1.Pod) {
+	if obj.Spec.Resources.Limits == nil || len(obj.Spec.Resources.Limits) > 0 {
```
No, we don't default limits when they are unset.

cici37 replied:
so it should be 
```suggestion
	if obj.Spec.Resources.Limits == nil || len(obj.Spec.Resources.Limits) == 0 {
```
Also seems like no tests covered for this

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4283,6 +4289,76 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+func validatePodResourceConsistency(spec *core.PodSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	v1PodSpec := &v1.PodSpec{}
```
Comment on why we are doing this?

thockin commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -429,3 +435,55 @@ func SetDefaults_HostPathVolumeSource(obj *v1.HostPathVolumeSource) {
+// while mainting compatability with the container-level specifications, as detailed
+// in KEP-2837: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2837-pod-level-resource-spec/README.md#proposed-validation--defaulting-rules
+func defaultPodResources(obj *v1.Pod) {
+	if obj.Spec.Resources.Limits == nil || len(obj.Spec.Resources.Limits) == 0 {
```
This can just be `len()` - calling that on a nil map will return 0

thockin commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -379,6 +379,603 @@ func testPodDefaults(t *testing.T, featuresEnabled bool) {
+		expectedPodSpec          v1.PodSpec
+	}{{
+		name:       "pod resources=unset, container resources=unset",
+		containers: []v1.Container{{Resources: v1.ResourceRequirements{}}, {Resources: v1.ResourceRequirements{}}},
```
hard to read :)

thockin reviewed (commented):
I didn't read everything but focused on APi related changes first.

msau42 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4283,6 +4289,76 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+	// requests requirements of a pod. Hence a local copy of Pod object is created from PodSpec to
+	// facilitate proper resource calculation without modifying PodRequests method.
+	podReqsFromContainers := resourcehelper.PodRequests(&v1.Pod{Spec: *v1PodSpec}, resourcehelper.PodResourcesOptions{
+		InPlacePodVerticalScalingEnabled: utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling),
```
I am not sure if we want to guard counting depending on if feature gates are enabled or not. Otherwise you may end up in a situation where an existing Pod could becomes invalid when you enable a feature gate.

ndixita replied:
This is highlighted in the KEP that a running pod will continue to run, but will become invalid on toggling the feature gate

msau42 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4283,6 +4289,76 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+
+		fldPath := reqPath.Key(resourceName.String())
+		if quantity.Cmp(podSpecRequests) > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath, podSpecRequests.String(), fmt.Sprintf("must be greater than or equal to effective %s container requests of %s", resourceName, quantity.String())))
```
What does effective container request mean? It does not appear to be a terminology we use in our docs.

yujuhong replied:
@ndixita may be we can add some comment to explain the meaning. This is not in the user documentation because it's used internally to decide how much resources is needed for the pod.

yujuhong replied:
Ah...I guess we'll have to document this more clearly since the error is user-facing. How about "aggregate container requests at each stage (init or non-init)".

thockin replied:
I think defining "aggregage container requests" makes sense - it's a better term than "effective" here, IMO.

ndixita replied:
modified, PTAL.

msau42 commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -429,3 +435,55 @@ func SetDefaults_HostPathVolumeSource(obj *v1.HostPathVolumeSource) {
 	}
 }
+
+// defaultPodResources applies default values for pod-level requests, when pod-level
```
I'm having troiuble finding the issue link, but I believe we cannot default fields that are in the pod template, otherwise the workload controllers will detect that as a template update and do a rolling update. @thockin do you remember this issue?

ndixita replied:
cc: @liggitt  as well

ndixita replied:
Thanks @yujuhong! yeah which is why I didn't add the defaulting logic in Pod Spec even though the new field is getting added in PodSpec. marking this as resolved.

AxeZhan commented on pkg/scheduler/framework/types.go:
```diff
@@ -1049,19 +1049,46 @@ func (n *NodeInfo) update(pod *v1.Pod, sign int64) {
+
+	// DefaultMilliCPURequest serves as the fallback value when both
+	// pod-level and container-level CPU requests are unspecified.
+	if _, exists := requests[v1.ResourceCPU]; !exists {
```
In this scenario:

1. Pod has pod-level mem request set, no cpu request set.
2. Pod has two containers, one with cpu request set, one not.

Then, we'll ignore cpu in L1063, and led us to a wrong non0Requests, no?

I guess we can get non0Requests  in L1086 first, and then replace them if pod-level cpu/memory set instead of using this `getNonMissingContainerRequests` ?

ndixita replied:
@AxeZhan: The defaulting logic will take care of propagating the requests at container level to pod-level if the request for a resource is not set at pod-level, but is set at container-level. So this won't be a problem. Let me try to add a comment and also an e2e test for the same.

tallclair commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -44,15 +45,80 @@ type PodResourcesOptions struct {
 func PodRequests(pod *v1.Pod, opts PodResourcesOptions) v1.ResourceList {
 	// attempt to reuse the maps if passed, or allocate otherwise
 	reqs := reuseOrClearResourceList(opts.Reuse)
+	effectiveContainersReqs := effectiveContainersRequests(pod, opts)
+	if !opts.UseContainerLevelResources && IsPodLevelRequestsSet(pod, opts.PodLevelResourcesEnabled) {
+		podRequests := v1.ResourceList{}
+		for resourceName, quantity := range pod.Spec.Resources.Requests {
+			if IsSupportedPodLevelResource(resourceName) {
+				podRequests[resourceName] = quantity
+			}
+		}
 
+		// Total pod resource requests are calculated using the effective container
+		// requests when resources are not specified at pod-level or when resources
+		// are not supported by PodLevelResource feature
+		for resourceName, quantity := range effectiveContainersReqs {
+			if _, exists := podRequests[resourceName]; !exists {
+				podRequests[resourceName] = quantity
+			}
+		}
+		addResourceList(reqs, podRequests)
+	} else {
+		addResourceList(reqs, effectiveContainersReqs)
+	}
```
I think this should reuse reqs, and then overwrite the values with the pod requests as needed. Otherwise it sort of defeats the purpose of avoiding allocating for reqs right? This also simplifies this method a lot.

```suggestion
	reqs := effectiveContainersRequests(pod, opts) // <-- do the clear/reuse in here
	if !opts.UseContainerLevelResources && IsPodLevelRequestsSet(pod, opts.PodLevelResourcesEnabled) {
		for resourceName, quantity := range pod.Spec.Resources.Requests {
			if IsSupportedPodLevelResource(resourceName) {
				reqs[resourceName] = quantity
			}
		}
	}
```

ndixita replied:
Thank you for catching this. In attempt to amend to the existing logic, my brain couldn't think of this solution! Extremely valuable!

ndixita committed:
e2e tests

Signed-off-by: ndixita <ndixita@google.com>

ndixita commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -429,3 +435,55 @@ func SetDefaults_HostPathVolumeSource(obj *v1.HostPathVolumeSource) {
+// pod-level limits.
+// This defaulting behavior ensures consistent resource accounting at the pod-level
+// while mainting compatability with the container-level specifications, as detailed
+// in KEP-2837: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2837-pod-level-resource-spec/README.md#proposed-validation--defaulting-rules
```
Fixed the comment for clarity. The default requests are applied only when limits are set explicitly + feature gate is enabled. Yes the default settings are validated in validation logic.

ndixita commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -429,3 +435,55 @@ func SetDefaults_HostPathVolumeSource(obj *v1.HostPathVolumeSource) {
+	}
+
+	podReqsFromContainers := resourcehelper.PodRequests(obj, resourcehelper.PodResourcesOptions{
+		InPlacePodVerticalScalingEnabled: utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling),
```
No, it is not dependent on the alpha feature.

ndixita commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4283,6 +4289,76 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+	// Pod-level requests >= effective requests of all containers in a pod.
+	for resourceName, quantity := range podReqsFromContainers {
+		podSpecRequests, exists := spec.Resources.Requests[core.ResourceName(resourceName.String())]
+		if !exists {
```
you are right here, this was a redundant check.

ndixita commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -44,15 +45,80 @@ type PodResourcesOptions struct {
+// If PodResourcesOptions is nil, then the requests are returned including pod overhead.
+// If the PodLevelResources feature is enabled AND the pod-level resources are set,
+// those pod-level values are used in calculating Pod Requests.
+// The computation is part of the API and must be reviewed as an API change.
```
I will send a follow up PR for this since this package has other methods that should trigger api review, but they exist in this file.

toVersus commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -379,6 +379,832 @@ func testPodDefaults(t *testing.T, featuresEnabled bool) {
+			}
+			p := v1.Pod{Spec: *spec.DeepCopy()}
+			corev1.SetDefaults_Pod(&p)
+			for i, container := range p.Spec.Containers {
```
Aren't we checking `limits` here? Same for pod resources. Enabling checks for `limits` will cause some test cases to fail.

ndixita replied:
the defaults are applied only to requests, hence checking only requests.

toVersus commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -379,6 +379,832 @@ func testPodDefaults(t *testing.T, featuresEnabled bool) {
+								"cpu":    resource.MustParse("1m"),
+								"memory": resource.MustParse("2Mi"),
+							},
+							Limits: v1.ResourceList{
```
Where did this limits value come from?

ndixita replied:
Fixed. This test worked because at container-level request = limit defaults are applied by k8s

ndixita commented:
/test all

toVersus commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -379,6 +379,828 @@ func testPodDefaults(t *testing.T, featuresEnabled bool) {
+				Limits: v1.ResourceList{
+					"cpu":     resource.MustParse("2m"),
+					"memory":  resource.MustParse("1Mi"),
+					"storage": resource.MustParse("1Mi"),
```
In Pod level resources, anything other than `cpu` and `memory` currently results in a validation error, so specifying `storage` as a podResources here seems a bit odd to me. Since `expectedPodSpec` does not include `storage` in pod level resources, this would cause an error if the unit tests were also checking the values of limits.

toVersus commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -838,6 +838,11 @@ func describePod(pod *corev1.Pod, events *corev1.EventList) (string, error) {
 		}
 
+		if pod.Spec.Resources != nil {
+			w.Write(LEVEL_0, "Resources:\n", "")
```
nit: I think the empty string for the third argument is also unnecessary.

```suggestion
			w.Write(LEVEL_0, "Resources:\n")
```

toVersus commented:
Enabling the `InPlacePodVerticalScalingEnabled` option in `PodValidationOptions` here should resolve the `TestValidatePodUpdate` unit test failure.
https://github.com/kubernetes/kubernetes/blob/0f4b24a4e2861995fae197a4feb6df3516854e9d/pkg/apis/core/validation/validation_test.go#L13702

Additionally, adding `Resources` to `osNeutralFields` should resolve the `TestValidateOSFields` unit test failure.

ndixita commented on pkg/scheduler/framework/types.go:
```diff
@@ -1049,19 +1049,46 @@ func (n *NodeInfo) update(pod *v1.Pod, sign int64) {
 	n.Generation = nextGeneration()
 }
 
+func getNonMissingContainerRequests(requests v1.ResourceList, podLevelResourcesSet bool) map[v1.ResourceName]resource.Quantity {
```
done

ndixita committed:
Limit Range changes to validate against Pod Level Resources

yujuhong commented on test/e2e/framework/pod/utils.go:
```diff
@@ -275,3 +277,18 @@ func FindContainerStatusInPod(pod *v1.Pod, containerName string) *v1.ContainerSt
 	}
 	return nil
 }
+
```
Add a comment for the function

yujuhong commented on test/e2e/common/node/pod_level_resources.go:
```diff
@@ -0,0 +1,319 @@
+	podLevelResourcesTests(f)
+})
+
+type ContainerInfo struct {
```
Is there a need to export the type? Same for the `ResourceInfo`

ndixita replied:
changing to un-exported type until new tests in common in test/e2e/node are added.

yujuhong commented on test/e2e/common/node/pod_level_resources.go:
```diff
@@ -0,0 +1,319 @@
+
+func VerifyPodCgroups(ctx context.Context, f *framework.Framework, pod *v1.Pod, expectedResources ResourceInfo) error {
+	ginkgo.GinkgoHelper()
+	if pod.Spec.Resources != nil {
```
Return early if pod resource spec is nil

yujuhong commented on test/e2e/common/node/pod_level_resources.go:
```diff
@@ -0,0 +1,319 @@
+	ginkgo.GinkgoHelper()
+	if pod.Spec.Resources != nil {
+		cmd := fmt.Sprintf("find /tmp -name '*%s*'", strings.ReplaceAll(string(pod.UID), "-", "_"))
+		framework.Logf("Namespace %s Pod %s container - finding Pod cgroup directory path: %q", cmd)
```
"container" seems redundant
s/Namespace %s Pod %s container/Namespace %s Pod %s

yujuhong commented on test/e2e/common/node/pod_level_resources.go:
```diff
@@ -0,0 +1,319 @@
+			return fmt.Errorf("encountered error while running command: %q, \nerr: %v \nstdErr: %q", cmd, err, stderr)
+		}
+
+		cpuWeightCgPath := fmt.Sprintf("%s/cpu.weight", podCgPath)
```
Instead of hard coding these, could you define `cpu.weight` and others (e.g., `cpu.max`,...) as constants properly? Then you can join them as needed later.

yujuhong commented on test/e2e/common/node/pod_level_resources.go:
```diff
@@ -0,0 +1,319 @@
+
+		cpuWeightCgPath := fmt.Sprintf("%s/cpu.weight", podCgPath)
+		expectedCPUShares := int64(kubecm.MilliCPUToShares(pod.Spec.Resources.Requests.Cpu().MilliValue()))
+		// convert cgroup v1 cpu.shares value to cgroup v2 cpu.weight value
```
The converting code are shared by `VerifyPodContainersCgroupValues` in `resize.go` as well. Consider factoring out to share.

yujuhong commented on test/e2e/common/node/pod_level_resources.go:
```diff
@@ -0,0 +1,319 @@
+		// convert cgroup v1 cpu.shares value to cgroup v2 cpu.weight value
+		// https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2#phase-1-convert-from-cgroups-v1-settings-to-v2
+		expectedCPUShares = int64(1 + ((expectedCPUShares-2)*9999)/262142)
+		err = e2epod.VerifyCgroupValue(f, pod, pod.Spec.Containers[0].Name, cpuWeightCgPath, strconv.FormatInt(expectedCPUShares, 10))
```
nit: 
```
if err := xxx; err != nil {
```
Same for others below

yujuhong commented on test/e2e/common/node/pod_level_resources.go:
```diff
@@ -0,0 +1,319 @@
+	}
+}
+
+func VerifyContainersCgroupLimits(f *framework.Framework, pod *v1.Pod) error {
```
no need to export the function?

ndixita replied:
changing until new tests are added in another package.

yujuhong commented on test/e2e/common/node/pod_level_resources.go:
```diff
@@ -0,0 +1,319 @@
+			err := VerifyPodCgroups(ctx, f, pod, *tc.podResources)
+			framework.ExpectNoError(err, "failed to verify pod's cgroup values: %v", err)
+
+			ginkgo.By("verifying containers cgroup limits are same as pod container's cgroup limits")
```
For clarity
s/same as pod container's cgroup limits/same as pod's cgroup limits when not set

yujuhong commented on test/e2e/common/node/pod_level_resources.go:
```diff
@@ -0,0 +1,319 @@
+	}
+}
+
+func VerifyContainersCgroupLimits(f *framework.Framework, pod *v1.Pod) error {
```
nit: instead of returning error on the first failure, you might want to aggregate the errors (or return a list of errors) to facilitate debugging.

yujuhong commented on test/e2e/feature/feature.go:
```diff
@@ -264,6 +264,9 @@ var (
 	// TODO: document the feature (owning SIG, when to use this feature for a test)
 	PodGarbageCollector = framework.WithFeature(framework.ValidFeatures.Add("PodGarbageCollector"))
 
+	// TODO: document the feature (owning SIG, when to use this feature for a test)
```
Why not just add the comment?

yujuhong commented on test/e2e/common/node/pod_level_resources.go:
```diff
@@ -0,0 +1,319 @@
+	return nil
+}
+
+func podLevelResourcesTests(f *framework.Framework) {
```
Could you also add a test to ensure that when the feature is enabled, but pod resources are not set, the container resources and cgroup values are not affected? If this is already covered by other test case, feel free to answer and resolve the comment.

yujuhong commented on pkg/apis/core/types.go:
```diff
@@ -3554,6 +3554,12 @@ type PodSpec struct {
 	ResourceClaims []PodResourceClaim
+	// Resources is the total amount of compute resources required by all
+	// containers in the pod. Currently, only CPU and Memory resources are
+	// supported.
```
I think it's validated in `validatePodResourceName`

yujuhong commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4030,6 +4030,17 @@ type PodSpec struct {
 	// +optional
 	ResourceClaims []PodResourceClaim `json:"resourceClaims,omitempty" patchStrategy:"merge,retainKeys" patchMergeKey:"name" protobuf:"bytes,39,rep,name=resourceClaims"`
+	// Resources is the total amount of compute resources required by all
+	// containers in the pod. Currently, only CPU and Memory resources are
```
The defaulting and interaction is non-trivial (as a result from making it more consistent and intuitive for individual cases). I think we can state the high-level principle and link to the detailed documentation (or KEP for now).

ndixita replied:
modified the comment. PTAL

yujuhong commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4058,6 +4059,10 @@ type PodValidationOptions struct {
 	AllowRelaxedDNSSearchValidation bool
 	// Allows zero value for Pod Lifecycle Sleep Action
 	AllowPodLifecycleSleepActionZeroValue bool
+	// Indicates whether pod-level resource feature is enabled or disabled.
```
feature name is wrong.

yujuhong commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5185,9 +5260,11 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
-		// resource values will result in validation failure.
-		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		// CPU and memory resource updates will result in validation failures in two cases:
+		// 1. When the feature-gate is disabled
```
nit: name the feature explicitly: "When InPlacePodVerticalScaling is disabled"

yujuhong commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5185,9 +5260,11 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
+		// 1. When the feature-gate is disabled
+		// 2. When using PodLevelResources, as it is incompatible with InPlacePodVerticalScaling
+		if opts.InPlacePodVerticalScalingEnabled &&
+			!(isPodLevelResourcesSet(oldPod) || isPodLevelResourcesSet(newPod)) {
```
`isPodLevelResourcesSet` checks two things: 1) whether the feature PodLevelResource sis enabled and 2) whether pod level resources are set. 

Q: Is it possible to have a case where pod level resources are set but the feature gate is disabled?

thockin replied:
Yeah, I don't think you want to check the gate here.  I think simply checking if there are pod resources set AT ALL means that resize should fail.  Beware the `resources: {}` case - is that "set" or "not set" ?

ndixita committed:
Resource Quota enforcement changes for Pod Level Resources

yujuhong commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -44,15 +45,70 @@ type PodResourcesOptions struct {
+	// UseContainerLevelResources controls whether pod resource calculation should use
+	// container-level resources to calculate PodRequests even when
+	// PodLevelResources feature is enabled and pod-level resources are set.
+	UseContainerLevelResources bool
```
I think setting this option `UseContainerLevelResources` caused the confusion from @haircommander in https://github.com/kubernetes/kubernetes/pull/128407/commits/56235c8d4f38bc2c1884e54b1c50122870fd3ab9#r1825874699. 

The source of the confusion probably comes from the fact that we reuse the same function for different purposes and simply set these boolean values to distinguish them. 

Some thoughts since this function is considered an API: 
 - Most components should not need to know when to set this option. It's not set by a feature gate and is not clear. 
 - Create another function for the use case where `UseContainerLevelResources` is always true. Components that know to call this will call this new function directly.

This likely is an overkill given that the list of fields in `PodResourcesOptions` is already very long with in-depth knowledge required to set them. It seems more like a utility function than an API in any way.

Tagging the API reviewers to see if they have any thought @thockin @cici37 
Otherwise, feel free to resolve this comment

ndixita replied:
I prefer adding a separate function AggregateContainerRequests. The only problem is that it needs some params from PodResourcesOptions. Right now AggregateContainerRequests accepts PodResourceOptions. Any suggestion for this?

Also, I wanted to move AggregateContainerRequests to an internal helper function, but I don't see any internal helpers being imported in staging/src/k8s.io. Is there a standard around this?

roycaihw commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4214,6 +4218,11 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
 	allErrs = append(allErrs, validateEphemeralContainers(spec.EphemeralContainers, spec.Containers, spec.InitContainers, vols, podClaimNames, fldPath.Child("ephemeralContainers"), opts, &spec.RestartPolicy, hostUsers)...)
+
+	if opts.PodLevelResourcesEnabled {
+		allErrs = append(allErrs, validatePodResources(spec, podClaimNames, fldPath.Child("resources"), opts)...)
```
IIUC `opts.PodLevelResourcesEnabled` is equal to `utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources) ` here.

@msau42 Is it possible for the field to be set when the feature gate is disabled? If so, we will need to reject create requests. Also does ratcheting validation apply here?

roycaihw replied:
@kerthcet Ah I missed it. Thanks for pointing that out!

thockin replied:
Right - it IS possible to arrive here with the gate disabled AND a value set, so we always use this pattern in validation.

ndixita commented:
/test all

ndixita commented:
/test all

AxeZhan commented:
lgtm from scheduling perspective.
/cc @sanposhiho @kerthcet @alculquicondor 
To approve scheduler changes.

kerthcet commented on pkg/scheduler/framework/types.go:
```diff
@@ -1052,19 +1052,73 @@ func (n *NodeInfo) update(pod *v1.Pod, sign int64) {
+// is included in the returned ResourceList.
+// Note that these default values are not set in the Pod object itself, they are only used
+// by the scheduler during node selection.
+func getNonMissingContainerRequests(requests v1.ResourceList, podLevelResourcesSet bool) v1.ResourceList {
```
@alculquicondor  We used to have the discussion about the default values for CPU and memory is a problem if we have multiple containers for one pod or multiple Pods in a node (I can't find the issue), maybe it's a good chance to have the default values for a Pod rather than each containers?

ndixita replied:
We discussed about the same in sig node; As much as we want to have the pod-level default values as a part of this PR, we are very late given the code freeze is in 2 days! I would like to keep the behavior as-is for now, and address this in next release.

alculquicondor replied:
I agree that it's too late to change at this point.

ndixita replied:
@alculquicondor please add a link to the discussion. I will add it to my notes, so that I can get to it in the next release.

ndixita replied:
No worries, added this as a TODO in my notes. Resolving the comment for now.

kerthcet replied:
Yeah, just a reminder, we can follow this later.

kerthcet commented on plugin/pkg/admission/limitranger/admission.go:
```diff
@@ -551,14 +552,24 @@ func PodValidateLimitFunc(limitRange *corev1.LimitRange, pod *api.Pod) error {
 func podRequests(pod *api.Pod, opts podResourcesOptions) api.ResourceList {
 	reqs := api.ResourceList{}
+	if opts.PodLevelResourcesEnabled && pod.Spec.Resources != nil {
+		addResourceList(reqs, pod.Spec.Resources.Requests)
```
So I guess this is because podspec resource is all we need to care about, e.g. we have considered the initContainer in the api server, so we can directly return here, right? Will the pod level resources be partly empty, like requests are set but limits are empty?

ndixita replied:
I modified the logic since I changed the same in the staging/src/k8s.io/component-helpers/resource/helpers.go. Podspec resoure is all we need to care about only for cpu and memory resources since only those are supported at pod-level.

kerthcet commented on plugin/pkg/admission/limitranger/admission_test.go:
```diff
@@ -450,13 +461,38 @@ func TestPodLimitFunc(t *testing.T) {
 	}
 	for i := range successCases {
 		test := successCases[i]
+
```
Can we refactor with t.Run() this time so the `featuregatetesting.SetFeatureGateDuringTest` works as expected. And we need to add this to the failure cases as well.

ndixita replied:
Refactored bare minimum for now. Added t.Run(); PTAL

kerthcet reviewed (commented):
Some comments left. Thanks!

yujuhong commented:
/assign @yujuhong

ndixita commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -429,3 +435,52 @@ func SetDefaults_HostPathVolumeSource(obj *v1.HostPathVolumeSource) {
+
+// defaultPodResources applies default values for pod-level requests, only when
+// pod-level limits are set, in following scenarios:
+// 1. When at least one container has requests set: The pod-level requests become
```
done!

ndixita commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -429,3 +435,55 @@ func SetDefaults_HostPathVolumeSource(obj *v1.HostPathVolumeSource) {
+// 2. When no containers have requests set: The pod-level requests become equal to
+// pod-level limits.
+// This defaulting behavior ensures consistent resource accounting at the pod-level
+// while mainting compatability with the container-level specifications, as detailed
```
done!

kerthcet replied:
I think not all addressed, the `mainting` unfixed.

kerthcet replied:
Just a nit, but kindly ping again incase you didn't notice this.

roycaihw commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4294,6 +4303,69 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+
+	reqPath := fldPath.Child("requests")
+	// resourcehelper.PodRequests method requires a Pod object to calculate the total
+	// requests requirements of a pod. Hence a local copy of Pod object is created from PodSpec to
```
> Hence a local copy of Pod object is created from PodSpec

I'm a little confused by this comment. Do you mean we are creating a copy of the resource list (i.e. `podReqsFromContainers`)? IIUC we are not creating a copy of the pod object.

ndixita replied:
fixed the language in the comment. I hope it makes sense now. PTAL.

ndixita commented on pkg/apis/core/types.go:
```diff
@@ -3554,6 +3554,12 @@ type PodSpec struct {
+	// containers in the pod. Currently, only CPU and Memory resources are
+	// supported.
+	// +featureGate=PodLevelResources
+	// +optional
```
fixed

fedebongio commented:
/remove-sig api-machinery

fedebongio commented:
/help

ndixita commented:
/retest

ndixita commented:
/test  pull-kubernetes-e2e-kind-ipv6

bart0sh commented:
/triage accepted

thockin commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -216,6 +217,11 @@ func SetDefaults_Pod(obj *v1.Pod) {
 		}
 	}
+
+	if resourcehelper.IsPodLevelResourcesSet(obj, utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources)) {
```
I think this warrants a comment.  To a naive reader it looks like: "If it is already set, then set it", which isn't REALLY what's happening.  Also, I don't think passing the gate down helps readability.  Maybe simpler:

```
if utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources) {
    defaultPodResources(obj)
}
```

and

```
func defaultPodResources(obj *v1.Pod) {
    // We only populate defaults when the pod-level resources are partly specified already.
    if !resourcehelper.IsPodLevelResourcesSet(obj) {
        return
    }
```

thockin commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -429,3 +435,53 @@ func SetDefaults_HostPathVolumeSource(obj *v1.HostPathVolumeSource) {
+// This defaulting behavior ensures consistent resource accounting at the pod-level
+// while mainting compatibility with the container-level specifications, as detailed
+// in KEP-2837: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2837-pod-level-resource-spec/README.md#proposed-validation--defaulting-rules
+func defaultPodResources(obj *v1.Pod) {
```
Maybe rename to `defaultPodResourceRequests()`

thockin reviewed (commented):
From commit "Adding the logic to set default pod-level request as following"

ndixita commented:
/test pull-kubernetes-verify

thockin commented on pkg/apis/core/helper/helpers.go:
```diff
@@ -186,7 +186,13 @@ func IsExtendedResourceName(name core.ResourceName) bool {
+
+// IsPrefixedNativeResource returns true if the resource name
+// has a kubernetes.io/ substring.
+func IsPrefixedNativeResource(name core.ResourceName) bool {
```
"Native" doesn't add any meaning here, and actually makes it a bit harder to understand - does it just mean "cpu" and "memory"?  No, it's really `HasKubePrefix` or something?

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4294,6 +4303,71 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+	// Convert the *core.PodSpec to *v1.PodSpec to satisfy the call to
+	// resourcehelper.PodRequests method, in the subsequent lines,
+	// which requires a *v1.Pod object (containing a *v1.PodSpec).
+	v1PodSpec := &v1.PodSpec{}
```
@jpbetz is this conversion likely to be problematic?  If we need to, we can reimplement `AggregateContainerRequests` against internal (maintaining two copies of these sorts of things is the #1 reason I think we should get rid of "internal")

@ndixita this came up in a different conversation that conversions CAN be expensive, and PodSpec is one of the larger resources.

jpbetz replied:
Yes, pods are write heavy and in the critical path of scale. This looks risky to me.  It's safest to assume that it is that it's problematic unless proven otherwise.  A scale test run that sig-scalability signs off on would be my bar.

I'd probably try to find some way to introduce a `resourcehelper.AggregateContainerRequests` variation that works on the internal type.

If this is not fixed now, I'd probably add "either scale test or optimize away conversion" to the beta graduation of this enhancements KEP.

> maintaining two copies of these sorts of things is the #1 reason I think we should get rid of "internal"

+100, I'd strongly prefer we pick a hub type and not maintain an extra internal type. That is, a type that all other types convert through and performs all the roles currently performed by the internal type.

ndixita replied:
So we can maintain a copy of this for internal type. I have 2 concerns:
1. I couldn't find an IFTTT example in the code that will help make sure all the copies of same code are updated together.
2. pull-kubernetes-verify https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128407/pull-kubernetes-verify/1853996449751109632 forbids  importing k8s.io/component-helpers/resource in  k8s.io/kubernetes/pkg/apis/core/v1: So does that mean we need 3 copies of the code?

ndixita replied:
@jpbetz @thockin ^

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4294,6 +4303,71 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+	})
+
+	// Pod-level requests >= effective requests of all containers in a pod.
+	for resourceName, quantity := range podReqsFromContainers {
```
super nits on naming:  I think this could be easier to read along the lines of:

```
ttlCtrReqs := resourcehelper.AggregateContainerRequests(...)

for resourceName, ctrReq := range ttlCtrReqs {
    key := resourceName.String()
    podReq := spec.Resources.Requests[core.ResourceName(key)]

    fldPath := reqPath.Key(key)
    if ctrReq.Cmp(podReq) > 0 {
        // error: sum of container requests > pod request
```

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4294,6 +4303,71 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+
+			if ctrLimit.Cmp(podSpecLimits) > 0 {
+				fldPath := fldPath.Child("containers").Index(i).Key(resourceName.String()).Child("limits")
+				allErrs = append(allErrs, field.Invalid(fldPath, ctrLimit.String(), fmt.Sprintf("must be less than or equal to pod %s limits of %s", resourceName, podSpecLimits.String())))
```
the resourcename is already in the field path, don't need it twice.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4294,6 +4303,71 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+
+		fldPath := reqPath.Key(resourceName.String())
+		if quantity.Cmp(podSpecRequests) > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath, podSpecRequests.String(), fmt.Sprintf("must be greater than or equal to effective %s container requests of %s", resourceName, quantity.String())))
```
the resourcename is already in the field path, don't need it twice.

thockin reviewed (commented):
commit "Adding the logic to validate pod-level resources as following"

ndixita commented:
/test all

ndixita commented:
/test all

ndixita commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -44,15 +45,66 @@ type PodResourcesOptions struct {
+// IsPodLevelResourcesSet check if PodLevelResources feature is enabled and
+// pod-level resources are set.
+func IsPodLevelResourcesSet(pod *v1.Pod, podLevelResourcesEnabled bool) bool {
+	return podLevelResourcesEnabled && pod.Spec.Resources != nil
```
changed it to:
return pod.Spec.Resources != nil &&
		(len(pod.Spec.Resources.Requests) > 0 ||
			len(pod.Spec.Resources.Limits) > 0)

It is confusing? Added a comment for clarity. Also wanted to remove dependency on feature gate since it will always be true in Beta and GA.

ndixita committed:
Fixing linter failures

ndixita commented:
/test pull-kubernetes-verify

kerthcet commented on pkg/scheduler/framework/types.go:
```diff
@@ -1052,19 +1052,72 @@ func (n *NodeInfo) update(pod *v1.Pod, sign int64) {
+		PodLevelResourcesEnabled: utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources),
+	})
+	podLevelResourcesSet := utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources) && resourcehelper.IsPodLevelResourcesSet(pod)
+	nonMissingContainerRequests := getNonMissingContainerRequests(requests, podLevelResourcesSet)
```
Do we  really need to calculate the missing requests? Shouldn't we apply the default values if not set? Then the defaults values will not change.

ndixita replied:
we still need this as the default values should be applied at container-level only of for that resource the requests are missing at pod-level.

kerthcet replied:
So once podLevelResources enabled, if we have a resource in pod level, but not in container level, we will not applying defaults , makes sense to me.

ndixita replied:
resolving this.

ndixita commented on plugin/pkg/admission/limitranger/admission.go:
```diff
@@ -591,9 +618,17 @@ func podRequests(pod *api.Pod) api.ResourceList {
+func podLimits(pod *api.Pod, opts podResourcesOptions) api.ResourceList {
 	limits := api.ResourceList{}
 
+	if opts.PodLevelResourcesEnabled && pod.Spec.Resources != nil && len(pod.Spec.Resources.Limits) > 0 {
```
fixed.

kerthcet commented:
/approve 
From sig-scheduling perspective unless others have other suggestions. cc @kubernetes/sig-scheduling-approvers

ndixita commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4203,6 +4204,11 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
 	allErrs = append(allErrs, validateInitContainers(spec.InitContainers, spec.Containers, vols, podClaimNames, gracePeriod, fldPath.Child("initContainers"), opts, &spec.RestartPolicy, hostUsers)...)
 	allErrs = append(allErrs, validateEphemeralContainers(spec.EphemeralContainers, spec.Containers, spec.InitContainers, vols, podClaimNames, fldPath.Child("ephemeralContainers"), opts, &spec.RestartPolicy, hostUsers)...)
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources) {
```
I added PodLevelResourcesEnabled in opts struct.

dom4ha commented:
/cc

dom4ha commented on pkg/scheduler/framework/plugins/noderesources/fit.go:
```diff
@@ -207,9 +213,11 @@ func NewFit(_ context.Context, plArgs runtime.Object, h framework.Handle, fts fe
 //
 // Result: CPU: 3, Memory: 3G
-func computePodResourceRequest(pod *v1.Pod) *preFilterState {
+func computePodResourceRequest(pod *v1.Pod, opts ResourceRequestsOptions) *preFilterState {
```
I'm make this and other function a method of the Fit plugin, but feel free to leave todo here for follow up instead.

ndixita replied:
Left a TODO

dom4ha replied:
I meant to make it a method, so to get rid of the opts param. Not sure if I understand the TODO

```suggestion
func (f *Fit) computePodResourceRequest(pod *v1.Pod) *preFilterState {
	// pod hasn't scheduled yet so we don't need to worry about InPlacePodVerticalScalingEnabled
	reqs := resource.PodRequests(pod, resource.PodResourcesOptions{
		PodLevelResourcesEnabled: f.enablePodLevelResources,
	})
	result := &preFilterState{}
	result.SetMaxResource(reqs)
	return result
```

dom4ha replied:
Yeah, I'm not sure if it's possible either, as this function is used in a few other places... so feel free to ignore my comment.

dom4ha commented on pkg/scheduler/framework/types.go:
```diff
@@ -1052,19 +1052,72 @@ func (n *NodeInfo) update(pod *v1.Pod, sign int64) {
+// is included in the returned ResourceList.
+// Note that these default values are not set in the Pod object itself, they are only used
+// by the scheduler during node selection.
+func getNonMissingContainerRequests(requests v1.ResourceList, podLevelResourcesSet bool) v1.ResourceList {
```
same here

ndixita replied:
fixed.

dom4ha reviewed (commented):
overall looks good

dims commented on test/e2e/common/node/pod_level_resources.go:
```diff
@@ -0,0 +1,332 @@
+	if pod.Spec.Resources == nil {
+		return nil
+	}
+	cmd := fmt.Sprintf("find /tmp -name '*%s*'", strings.ReplaceAll(string(pod.UID), "-", "_"))
```
Does this technique to mount /sys/fs/cgroup and find its own `cpu.weight` `cpu.max` etc work well across OS/Runtime/Cgroup versions etc?

dims replied:
cc @haircommander perhaps for crio?

haircommander commented on test/e2e/common/node/pod_level_resources.go:
```diff
@@ -0,0 +1,332 @@
+	if pod.Spec.Resources == nil {
+		return nil
+	}
+	cmd := fmt.Sprintf("find /tmp -name '*%s*'", strings.ReplaceAll(string(pod.UID), "-", "_"))
```
I found it really confusing seeing /tmp here. can you use a different path in the container, and also maybe pass around a variable for the location (so it's not hardcoded twice?)

aojea commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-e2e-gci-gce-alpha-enabled-default
/test pull-kubernetes-node-e2e-alpha-ec2
/test pull-kubernetes-node-e2e-containerd-alpha-features


Please execute jobs that exercise the code path before merge to avoid breaking existing jobs

Thanks

ndixita commented on pkg/scheduler/framework/plugins/noderesources/fit.go:
```diff
@@ -380,7 +388,7 @@ func (f *Fit) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldO
 	}
 	// The pod will fit, but since there was no increase in available resources, the change won't make the pod schedulable.
-	if !haveAnyRequestedResourcesIncreased(pod, originalNode, modifiedNode) {
+	if !haveAnyRequestedResourcesIncreased(pod, originalNode, modifiedNode, ResourceRequestsOptions{EnablePodLevelResources: f.enableInPlacePodVerticalScaling}) {
```
fixed.

ndixita commented on plugin/pkg/admission/limitranger/admission.go:
```diff
@@ -582,6 +597,18 @@ func podRequests(pod *api.Pod) api.ResourceList {
+			if isSupportedPodLevelResource(resourceName) {
+				reqs[resourceName] = quantity
+			}
+		}
```
added a comment instead for clarity.

ndixita commented on pkg/scheduler/framework/types.go:
```diff
@@ -1052,19 +1052,72 @@ func (n *NodeInfo) update(pod *v1.Pod, sign int64) {
+		UseStatusResources:       utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling),
+		PodLevelResourcesEnabled: utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources),
+	})
+	podLevelResourcesSet := utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources) && resourcehelper.IsPodLevelResourcesSet(pod)
```
I am going move all the helper methods to an internal package as the one of the verifiers is failing to import an external API with following error:

```Specifically #2 - validating imports for 11 packages
2024/11/06 03:39:13 -- found forbidden imports for [k8s.io/kubernetes/pkg/apis/core/v1](http://k8s.io/kubernetes/pkg/apis/core/v1): Will we have to maintain 3 copies of methods in staging/src/[k8s.io/component-helpers/resource/helpers.go](http://k8s.io/component-helpers/resource/helpers.go)
```

So if I have an internal package, I will use it for everywhere internally. Discussing with some folks before I make this change.

alculquicondor commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -46,15 +47,68 @@ type PodResourcesOptions struct {
-// as an API change.
+var supportedPodLevelResources = sets.New(v1.ResourceCPU, v1.ResourceMemory)
+
+func GetSupportedPodLevelResources() sets.Set[v1.ResourceName] {
```
```suggestion
func SupportedPodLevelResources() sets.Set[v1.ResourceName] {
```

alculquicondor commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -46,15 +47,68 @@ type PodResourcesOptions struct {
 	NonMissingContainerRequests v1.ResourceList
+	// PodLevelResourcesEnabled indicates that the pod-level resources feature gate
+	// is enabled.
+	PodLevelResourcesEnabled bool
```
maybe the name could be more similar to `ExcludeOverhead`: `ExcludePodLevelResources`

dom4ha commented:
/lgtm

ndixita committed:
Allowlisting k8s.io/components-helpers/resource import in ./pkg/apis/core

alculquicondor commented:
/approve
for scheduling

/label api-review

ndixita commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features

ndixita commented:
/test all

ndixita commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-e2e-gci-gce-alpha-enabled-default
/test pull-kubernetes-node-e2e-alpha-ec2
/test pull-kubernetes-node-e2e-containerd-alpha-features

kannon92 commented:
Can we add a dedicated test job for this?

I see that we added an alpha feature but I'm not sure if we have a test-infra job for this.

ndixita commented:
/test all
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-e2e-gci-gce-alpha-enabled-default
/test pull-kubernetes-node-e2e-alpha-ec2
/test pull-kubernetes-node-e2e-containerd-alpha-features

yujuhong commented on test/e2e/feature/feature.go:
```diff
@@ -264,6 +264,13 @@ var (
+	// TODO: document the feature (when to use this feature for a test)
+	PodLevelResources = framework.WithFeature(framework.ValidFeatures.Add("PodLevelResources"))
+
+	// TODO: document the feature (owning SIG, when to use this feature for a test)
```
Need to rebase to drop this?

ndixita commented:
/test pull-kubernetes-node-kubelet-serial-podresources

thockin reviewed (commented):
There's a TON going on in here, I hope I caught asll the places I wanted to look.

yujuhong commented:
The kubelet package change looks good overall.

ndixita commented:
/test pull-kubernetes-node-kubelet-serial-podresources

kannon92 commented:
/test

kannon92 commented:
/test pull-kubernetes-node-kubelet-serial-podresources

ndixita commented:
/retest-required

ndixita commented:
/test pull-kubernetes-node-kubelet-serial-podresources

ndixita commented:
@yujuhong @thockin cleaned up most parts of the code, added e2e for feature enabled + no pod resources case, and fixed other comments. Please let me know if you I'd need an extension.

ndixita commented:
/test pull-kubernetes-node-e2e-containerd

knabben commented:
Hey @ndixita @kannon92 
⚠️ Do we still intend to merge this for v1.32? Just a reminder that the code freeze is starting 02:00 UTC Friday November 8th 2024 (a little less than 1 week from now). Please make sure the PR has both lgtm and approved labels before the code freeze. Thanks!

kannon92 commented on pkg/api/pod/util.go:
```diff
@@ -386,6 +386,8 @@ func GetValidationOptionsFromPodSpecAndMeta(podSpec, oldPodSpec *api.PodSpec, po
 		AllowNamespacedSysctlsForHostNetAndHostIPC:        false,
 		AllowNonLocalProjectedTokenPath:                   false,
 		AllowPodLifecycleSleepActionZeroValue:             utilfeature.DefaultFeatureGate.Enabled(features.PodLifecycleSleepActionAllowZero),
+		InPlacePodVerticalScalingEnabled:                  utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling),
```
This doesn't seem related to your feature.

ndixita replied:
fixed

kannon92 commented on pkg/api/pod/util.go:
```diff
@@ -675,6 +677,11 @@ func dropDisabledFields(
 		}
 	}
 
+	// If the feature is disabled and not in use, drop Resources at the pod-level from PodSpec.
```
nit. Can we move this to a separate function like we have below in 685-687.

ndixita replied:
done

kannon92 replied:
I don't see a change here.

ndixita commented:
pull-kubernetes-e2e-gce-cos-alpha-features job failures don't seem related. It is flaky/broken https://prow.k8s.io/job-history/gs/kubernetes-ci-logs/pr-logs/directory/pull-kubernetes-e2e-gce-cos-alpha-features. 

@aojea is helping me check

aojea commented:
having all of these failing is a strong signal 
![image](https://github.com/user-attachments/assets/21320c11-f5cc-4b2d-bbd4-200ced84b5aa)


but let me check in some of the logs, in this cases please link the comment to the job that failed exactly , per example this


https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128407/pull-kubernetes-e2e-gce-cos-alpha-features/1854456454581325824

![image](https://github.com/user-attachments/assets/1503dea0-859f-415f-8aeb-184b0fb62da9)


The ClusterBundleThing is here ClusterTrustBundle we can see the rest are green, we can assume this is correct, that was on commit https://github.com/kubernetes/kubernetes/commit/2930d7ed294db26770c6e8f61fcb093d94159a8f

The kind jobs and ec2 failed in other commit https://github.com/kubernetes/kubernetes/commit/e11ac5da83ece9b3c37041aad2dc770db48df65e

Those have more errors https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128407/pull-kubernetes-e2e-kind-alpha-features/1854281568235819008

and seems to show that it is failing constantly in this PR https://prow.k8s.io/job-history/gs/kubernetes-ci-logs/pr-logs/directory/pull-kubernetes-e2e-kind-alpha-features ,

that may be a problem

kannon92 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4294,6 +4302,73 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+	aggrContainerReqs := resourcehelper.AggregateContainerRequests(&v1.Pod{Spec: *v1PodSpec}, resourcehelper.PodResourcesOptions{})
+
+	// Pod-level requests >= aggregate requests of all containers in a pod.
+	for resourceName, ctrReqs := range aggrContainerReqs {
```
You say that we are limiting this feature to cpu and memory but I don't really see any kind of enforcement in this code.

Should we block folks from setting anything other than cpu/memory for pod resources?

ndixita replied:
PTAL: https://github.com/kubernetes/kubernetes/pull/128407#discussion_r1833133704

kannon92 replied:
That sounds good. It may be worth having a few unit tests for validation that test what error folks get if they have an unsupported value.

ndixita replied:
I have the unit tests in place. Although they validation that the unsupported resources throw an error. They don't check the exact error message. Was trying to cater to how the existing tests are written :/

I will send a follow up PR for cleanup so that these tests are easier to identify. Marking the comment as resolved. Feel free to re-open it.

ndixita commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4294,6 +4302,73 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+
+	// ValidateResourceRequirements checks if resource names and quantities are
+	// valid, and requests are less than limits.
+	allErrs = append(allErrs, ValidatePodResourceRequirements(spec.Resources, podClaimNames, fldPath, opts)...)
```
@kannon92  ValidatePodResourceRequirements -> validatePodResourceName:This is where the resource names are validated. The pod-level resources consistency is checked after that

ndixita commented:
/test pull-kubernetes-e2e-kind-alpha-features

ndixita commented on pkg/apis/core/helper/qos/qos.go:
```diff
@@ -48,54 +89,44 @@ func GetPodQOS(pod *core.Pod) core.PodQOSClass {
-					requests[name] = delta
-				}
-			}
+	// When pod-level resources are specified, we use them to determine QoS class.
```
I haven't refactored the code much here; just moved the for loop code in a helper method to re-use.

ndixita commented on pkg/apis/core/helper/qos/qos.go:
```diff
@@ -113,11 +113,39 @@ func ComputePodQOS(pod *core.Pod) core.PodQOSClass {
 		allContainers = append(allContainers, pod.Spec.InitContainers...)
```
@kannon92 I want to revert this back :|
Doesn't help much with the diff in the PR. Also, there's test coverage for this package. Let me know your thoughts

kannon92 replied:
My main concern is that we want to isolate this change to just your feature. If you sprinkle in refactors and cleanups it can cause problems.

I think your PR is an improvement but it is also mixed in with a pretty large feature.

ndixita commented:
/test pull-kubernetes-e2e-kind-alpha-features

thockin commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -429,3 +435,56 @@ func SetDefaults_HostPathVolumeSource(obj *v1.HostPathVolumeSource) {
+	}
+
+	var reqs v1.ResourceList
+	reqs = obj.Spec.Resources.Requests
```
super nitpick, but since this code is SO important and is subtle:

1) Consider renaming `reqs` to `podReqs`
2) Consider renaming `podReqsFromContainers` to `aggCtrReqs` or `effPodReqs` or `sumCtrReqs`
3) Consider renaming `value` to `podLim` or `lim`

Thus, things like `reqs[key] = value.DeepCopy()` -> `podReqs[key] = podLim.DeepCopy()`

cici37 reviewed (commented):
Did another pass only focused on API related changes. :)

thockin commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -216,6 +217,11 @@ func SetDefaults_Pod(obj *v1.Pod) {
 		}
 	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources) {
```
Maybe put this higher up, right after "set requests to limits if requests are not specified, but limits are" -- this HAS TO run _after_ that (to get all container requests) and having them far apart makes it less obvious.  I would also add a comment about per-container must come first.

ndixita replied:
moved and added comment

thockin replied:
I don't see it moved?  I thought ~184 is wherre it should go

thockin commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -46,15 +47,71 @@ type PodResourcesOptions struct {
+		return false
+	}
+
+	return (len(pod.Spec.Resources.Requests) + len(pod.Spec.Resources.Limits)) > 0
```
As above, this will return true if:

```
   resources:
      limits:
        example.com/foobar: "0"
```

Is that OK?

ndixita replied:
Added the check for supported keys; however not checking is quantity is 0

thockin replied:
That's OK.  I used 0 to make a fake resource actually work.

thockin commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -46,15 +47,71 @@ type PodResourcesOptions struct {
+// IsPodLevelRequestsSet checks if pod-level requests are set. It returns true if
+// Requests map is non-empty.
+func IsPodLevelRequestsSet(pod *v1.Pod) bool {
+	return pod.Spec.Resources != nil && len(pod.Spec.Resources.Requests) > 0
```
same

ndixita replied:
fixed, need to dedup code between this and IsPodLevelResources() later

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4294,6 +4302,73 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+
+	allErrs := field.ErrorList{}
+
+	// ValidateResourceRequirements checks if resource names and quantities are
```
comment and code disagree on name

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6762,8 +6863,16 @@ func validateBasicResource(quantity resource.Quantity, fldPath *field.Path) fiel
+
 // Validates resource requirement spec.
-func ValidateResourceRequirements(requirements *core.ResourceRequirements, podClaimNames sets.Set[string], fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
+func ValidateResourceRequirements(requirements *core.ResourceRequirements, resourceNameFn func(core.ResourceName, *field.Path) field.ErrorList, podClaimNames sets.Set[string], fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
```
super nit, but this doesn't seem to need to be public (or else it would be a breaking change)

ndixita replied:
I thought so too. It was a public method before my change so kept it as-is.

thockin replied:
The new functions are better as a "public" API (even just within k/k) so I think it can be demoted.  More local is more better! :)

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5284,6 +5359,11 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 	return allErrs
 }
 
+func isPodLevelResourcesSet(pod *core.Pod) bool {
```
Really really minor OK to disregard, but I had to go search for callers to make sure this didn't need to handle unsupported keys.  

This could be a func literal at the callsite, to emphasize that it is HYPER local.  

```
    isPodLevelResourcesSet := func(pod *core.Pod) bool {
        return pod.Spec.Resources != nil &&
            len(pod.Spec.Resources.Requests) + len(pod.Spec.Resources.Limits) > 0
    }
    if isPodLevelResourcesSet(oldPod) || isPodLevelResourcesSet(newPod) {
        return...
```

thockin commented on pkg/scheduler/framework/types.go:
```diff
@@ -1052,19 +1052,74 @@ func (n *NodeInfo) update(pod *v1.Pod, sign int64) {
-		},
-	})
-
+	isPodLevelResourcesSet := utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources) && resourcehelper.IsPodLevelRequestsSet(pod)
```
it's these callsites I worry about wrt other keys, as example

ndixita replied:
fixed as the helper method is fixed

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4294,6 +4302,73 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+	if spec.Resources == nil {
+		return nil
+	}
+
```
I think you also want to ensure that `claims` is not set (`field.Forbidden()`) ?

thockin reviewed (changes requested):
I went thru API with an eye towards corner cases and how bad input could look.

ndixita commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5498,6 +5578,11 @@ func ValidatePodResize(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 	allErrs = append(allErrs, validatePodMetadataAndSpec(newPod, opts)...)
 
+	// pods with pod-level resources cannot be resized
+	if isPodLevelResourcesSet(oldPod) || isPodLevelResourcesSet(newPod) {
```
yes, since this is updating the created pod. If the pod was created with resources when feature gate was enabled, and then updated when feature gate is disabled, this should still fail.

cici37 replied:
I think as long as the feature gate is disabled, the validation should ignore the field and pass same as previous behavior. Update should not fail if the field is set but feature gate is disabled.
/cc @thockin for verification.

ndixita replied:
I should clarified this in my previous comment. We don't want to support in place pod resize with pod-level resources. Hence failing the validation here.

ndixita committed:
keeping the qos code as-is for the existing case when pod-level resources are not set

Signed-off-by: ndixita <ndixita@google.com>

ndixita commented on pkg/api/pod/util.go:
```diff
@@ -1116,6 +1121,12 @@ func supplementalGroupsPolicyInUse(podSpec *api.PodSpec) bool {
+// podLevelResourcesInUse returns true if pod-spec is non-nil and Resources field at
+// pod-level is set.
+func podLevelResourcesInUse(podSpec *api.PodSpec) bool {
+	return podSpec != nil && podSpec.Resources != nil
```
fixed and added the test

ndixita commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4080,6 +4080,20 @@ type PodSpec struct {
 	ResourceClaims []PodResourceClaim `json:"resourceClaims,omitempty" patchStrategy:"merge,retainKeys" patchMergeKey:"name" protobuf:"bytes,39,rep,name=resourceClaims"`
+	// Resources is the total amount of CPU and Memory resources required by all
+	// containers in the pod. Currently, only "cpu" and "memory" resource names
+	// are supported.
```
Fixed

ndixita commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4294,6 +4302,73 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+			}
+
+			if ctrLimit.Cmp(podSpecLimits) > 0 {
+				fldPath := fldPath.Child("containers").Index(i).Key(resourceName.String()).Child("limits")
```
intentional as this is checking for container[i] and erroring out for the same.

thockin commented:
/approve API

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go:
```diff
@@ -167,13 +167,14 @@ func TestGenerateLinuxContainerConfigResources(t *testing.T) {
 	}{
 		{
 			name: "Request 128M/1C, Limit 256M/3C",
-			podResources: v1.ResourceRequirements{
```
Did you replace the test cases with the new ones? Why not keeping the originals too?

ndixita replied:
didn't remove the tests; the test had a field named "podResources" changed to to containerResourced and added tests with podResources relevant for my change

yujuhong replied:
NVM. The field just got renamed.

ndixita commented:
/test pull-kubernetes-node-kubelet-serial-podresources

yujuhong commented on test/e2e/common/node/pod_level_resources.go:
```diff
@@ -0,0 +1,409 @@
+}
+
+func verifyContainersCgroupLimits(f *framework.Framework, pod *v1.Pod) error {
+	// var err error
```
remove the stray comment

mrunalp commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -438,3 +446,56 @@ func SetDefaults_PodLogOptions(obj *v1.PodLogOptions) {
+// defaultPodRequests applies default values for pod-level requests, only when
+// pod-level limits are set, in following scenarios:
+// 1. When at least one container (regular, init or sidecar) has requests set:
+// The pod-level requests become equal to the effective requests of all containers
```
nit: I think it will be good to copy relevant rows from the table we have in the KEP as here.

ndixita replied:
Ack'ing for next release

ndixita commented:
/test pull-kubernetes-unit

ndixita commented:
/test pull-kubernetes-node-kubelet-serial-podresources

ndixita commented:
/test pull-kubernetes-node-kubelet-serial-podresources

mrunalp commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -46,15 +47,101 @@ type PodResourcesOptions struct {
+}
+
+// AggregateContainerRequests computes the total resource requests of all the containers
+// in a pod. This computation folows the formula defined in the KEP for sidecar
```
nit: follows

ndixita replied:
Fixed

mrunalp commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -46,15 +47,101 @@ type PodResourcesOptions struct {
+
+// AggregateContainerRequests computes the total resource requests of all the containers
+// in a pod. This computation folows the formula defined in the KEP for sidecar
+// containers. See https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/753-sidecar-containers#resources-calculation-for-scheduling-and-pod-admission
```
could consider inlining that section from the KEP here.

ndixita replied:
could be confusing since the KEP uses different terminology. 

"Max ( Max( each InitContainerUse ) , Sum(Sidecar Containers) + Sum(each ContainerUse) ) + pod overhead" from the KEP.
 
I myself don't understand it :')

ndixita commented:
/test pull-kubernetes-node-kubelet-serial-podresources

ndixita commented:
/test all

mrunalp commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4282,6 +4290,77 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+	// calculation without modifying AggregateContainerRequests method.
+	aggrContainerReqs := resourcehelper.AggregateContainerRequests(&v1.Pod{Spec: *v1PodSpec}, resourcehelper.PodResourcesOptions{})
+
+	// Pod-level requests >= aggregate requests of all containers in a pod.
```
nit: requests must be >= ..

ndixita replied:
Fixed

mrunalp commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4282,6 +4290,77 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
+		}
+	}
+
+	// Individual Container limits <= Pod-level limits.
```
nit: must be <=

ndixita replied:
Fixed

ndixita commented:
/test pull-kubernetes-node-kubelet-serial-podresources

ndixita commented:
/test all

mrunalp reviewed (commented):
Can we add a description to the commit: `QOS changes for Pod Level resources`?

yujuhong commented:
/lgtm

Still need to wait for the tests to pass

ndixita commented:
/test pull-kubernetes-e2e-kind

ndixita commented:
/test pull-kubernetes-verify

ndixita commented:
/test pull-kubernetes-node-kubelet-serial-podresources

yujuhong commented:
/lgtm

pacoxu commented:
/lgtm
after a typo fix

ndixita commented:
/test pull-kubernetes-node-kubelet-serial-podresources

ndixita commented:
/retest-required

ndixita commented:
/retest-required

ndixita commented:
The tests timing out in pull-kubernetes-e2e-gce-cos-alpha-features are not realated to this PR. They are related to InPlaceVerticalScaling feature

![image](https://github.com/user-attachments/assets/72895b4b-8719-46c2-9311-44613449e00e)
